source/connection-monitoring-and-pooling/tests/README.rst

```diff
@@ -0,0 +1,165 @@
+.. role:: javascript(code)
+  :language: javascript
+
+========================================
+Connection Monitoring and Pooling (CMaP)
+========================================
+
+.. contents::
+
+--------
+
+Introduction
+============
+
+The YAML and JSON files in this directory are platform-independent tests that
+drivers can use to prove their conformance to the Connection Monitoring and Pooling (CPAM) Spec.
+
+Several prose tests, which are not easily expressed in YAML, are also presented
+in this file. Those tests will need to be manually implemented by each driver.
+
+Common Test Format
+================
+
+Each YAML file has the following keys:
+
+- ``version``: A version number indicating the expected format of the spec tests (current version = 1)
+- ``style``: A string indicating what style of tests this file contains. Currently ``unit`` is the only valid value
+- ``tests``: An array of tests that are to be run independently of each other
+
+Unit Test Format:
+=================
+
+All Unit Tests have some of the following fields:
+
+- ``description``: A text description of what the test is meant to assert
+- ``operations``: A list of operations to perform. Each operation has some of the following fields:
+
+  - ``command``: A string describing which command to issue.
+  - ``args``: An array of arguments to pass to the command
+  - ``thread``: The name of the thread in which to run this command. If not specified, runs in the default thread
+  - ``object``: For some command, specifies the object on which to run the command
+  - ``returnTo``: For some commands, specifies what to name the return value of the command
+
+- ``error``: Indicates that the main thread is expected to error during this test. Subfields (like ``message``) indicate what error is expected
```

Comment:
Rather than saying "subfields like `message`", I think it would be better to actually document the error subfield in the same way that subfields for each `operation` are. It's not clear to me from the current wording whether `message` is required on all `error` fields of if there are any other possible subfields besides `message`, and I think it would be more convenient for implementers not to have to search through all of the tests to determine this.

Reply:
Like with my comment about `event` below, this is specifically pinned to how events are defined in the spec, and I'd like to avoid duplicating information.

Reply:
I think the issue is that the spec doesn't define what the errors look like in YAML, just what they look like as an abstract driver interface. For someone implementing the tests, they're going to have to go through every instance of `error` in the YAML files to figure out what they need to read in (as most languages cannot convert YAML/JSON to their native object type without knowing the structure of the object first), so I feel like it would be much more convenient to just list what the field names and value types are here.